File: clients/typescript/test/live.test.ts
/**
 * Integration tests against a live ZydecoDB server. Set ZYDECODB_TEST_HOST /
 * ZYDECODB_TEST_PORT (and optionally ZYDECODB_TEST_API_KEY) to point at a
 * running server; the suite is skipped when the server is unreachable, so a
 * plain `npm test` stays green offline.
 */
import assert from "node:assert/strict";
import net from "node:net";
import { test } from "node:test";

import { AuthError, Client, ConflictError, type Document } from "../src/index.ts";

const HOST = process.env.ZYDECODB_TEST_HOST ?? "127.0.0.1";
const PORT = Number(process.env.ZYDECODB_TEST_PORT ?? "9470");
const API_KEY = process.env.ZYDECODB_TEST_API_KEY ?? undefined;

function serverUp(): Promise<boolean> {
  return new Promise((resolve) => {
    const socket = net.connect({ host: HOST, port: PORT });
    const done = (up: boolean): void => {
      socket.destroy();
      resolve(up);
    };
    socket.setTimeout(1000);
    socket.once("connect", () => done(true));
    socket.once("timeout", () => done(false));
    socket.once("error", () => done(false));
  });
}

const skip = !(await serverUp());
const uniqueCollection = (): string => `tstest_${Date.now()}_${Math.floor(Math.random() * 1e6)}`;

function newClient(): Client {
  return new Client(`${HOST}:${PORT}`, API_KEY ? { apiKey: API_KEY } : {});
}

test("ping", { skip }, async () => {
  const db = newClient();
  try {
    await db.ping();
  } finally {
    db.close();
  }
});

test("raw kv", { skip }, async () => {
  const db = newClient();
  try {
    const key = Buffer.from(`testkv_${Date.now()}`);
    const val = Buffer.from("hello kv");

    // Get missing
    let res = await db.get(key);
    assert.equal(res, null);

    // Put
    const seq = await db.put(key, val, 0);
    assert.ok(seq > 0n);

    // Get
    res = await db.get(key);
    assert.ok(res);
    assert.equal(res.toString(), "hello kv");

    // Delete
    let existed = await db.delete(key);
    assert.equal(existed, true);
    existed = await db.delete(key);
    assert.equal(existed, false);

    // Put with TTL (expired)
    const expired = Date.now() - 3600000;
    await db.put(key, val, expired);
    res = await db.get(key);
    assert.equal(res, null);
  } finally {
    db.close();
  }
});

test("insert / find / update / delete", { skip }, async () => {
  const db = newClient();
  const coll = db.collection(uniqueCollection());
  try {
    await coll.createIndex(["age"]);
    const ids = await coll.insertMany([
      { name: "Ada", age: 30, city: "London" },
      { name: "Bo", age: 25, city: "NOLA" },
      { name: "Cy", age: 40, city: "NOLA" },
    ]);
    assert.equal(ids.length, 3);

    const got = await coll.find({ age: { $gte: 30 } }, { sort: [{ field: "age", ascending: true }] });
    assert.deepEqual(
      got.map((d: Document) => d.name),
      ["Ada", "Cy"],
    );

    const res = await coll.updateOne({ name: "Bo" }, { $inc: { age: 10 } });
    assert.equal(res.matched, 1);
    assert.equal(res.modified, 1);

    assert.equal(await coll.countDocuments(), 3);
    const cities = (await coll.distinct("city")) as string[];
    assert.deepEqual(cities.slice().sort(), ["London", "NOLA"]);

    assert.equal(await coll.deleteMany({ city: "NOLA" }), 2);
    assert.equal(await coll.countDocuments(), 1);
  } finally {
    db.close();
  }
});

test("unique index conflict", { skip }, async () => {
  const db = newClient();
  const coll = db.collection(uniqueCollection());
  try {
    await coll.createIndex(["email"], true);
    await coll.insertOne({ email: "a@b.com" });
    await assert.rejects(() => coll.insertOne({ email: "a@b.com" }), ConflictError);
  } finally {
    db.close();
  }
});

test("upsert setOnInsert", { skip }, async () => {
  const db = newClient();
  const coll = db.collection(uniqueCollection());
  try {
    const miss = await coll.updateOne(
      { email: "soi@example.com" },
      { $set: { email: "soi@example.com", n: 1 }, $setOnInsert: { created: true } },
      false,
      true,
    );
    assert.equal(miss.matched, 0);
    assert.equal(miss.modified, 0);
    assert.ok(miss.upserted_id);

    let doc = await coll.findOne({ email: "soi@example.com" });
    assert.equal(doc?.created, true);
    assert.equal(doc?.n, 1);

    const hit = await coll.updateOne(
      { email: "soi@example.com" },
      { $set: { n: 2 }, $setOnInsert: { created: false, extra: 1 } },
      false,
      true,
    );
    assert.equal(hit.matched, 1);
    assert.equal(hit.modified, 1);
    assert.equal(hit.upserted_id, undefined);

    doc = await coll.findOne({ email: "soi@example.com" });
    assert.equal(doc?.n, 2);
    assert.equal(doc?.created, true);
    assert.equal(doc?.extra, undefined);
  } finally {
    db.close();
  }
});

test("optimistic concurrency", { skip }, async () => {
  const db = newClient();
  const coll = db.collection(uniqueCollection());
  try {
    const id = await coll.insertOne({ n: 1 });
    const got = await coll.getWithRevision(id);
    assert.ok(got);
    assert.equal(got!.doc.n, 1);
    assert.ok(got!.revision > 0n);
    const newRev = await coll.replaceOneIfMatch(id, { n: 2 }, got!.revision);
    assert.ok(newRev > got!.revision);
    await assert.rejects(
      () => coll.replaceOneIfMatch(id, { n: 3 }, got!.revision),
      ConflictError,
    );
    const after = await coll.updateByIdIfMatch(id, { $inc: { n: 1 } }, newRev);
    assert.ok(after > newRev);
  } finally {
    db.close();
  }
});

test("bounded transaction", { skip }, async () => {
  const db = newClient();
  const collName = uniqueCollection();
  const coll = db.collection(collName);
  try {
    await coll.insertOne({ _seed: true });
    const key = Buffer.from(`tx-ts-${collName}`);
    const { seq } = await db.withTransaction(async (tx) => {
      await tx.put(key, Buffer.from("v1"));
      await tx.putDocument(collName, "u1", { n: 1 });
      const got = await tx.get(key);
      assert.ok(got);
      assert.equal(got!.toString("utf8"), "v1");
    });
    assert.ok(seq > 0n);
    const committed = await db.get(key);
    assert.equal(committed?.toString("utf8"), "v1");
  } finally {
    db.close();
  }
});

/** Resolve to `fallback` if `p` has not settled within `ms`. Clears its timer. */
function settleWithin<T, F>(p: Promise<T>, ms: number, fallback: F): Promise<T | F> {
  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => resolve(fallback), ms);
    p.then(
      (v) => {
        clearTimeout(timer);
        resolve(v);
      },
      (e) => {
        clearTimeout(timer);
        reject(e);
      },
    );
  });
}

test("watch idle timeout is decoupled from request timeout", { skip }, async (t) => {
  // A 1s request timeout must not apply to the dedicated Watch connection: an
  // idle stream only carries a server heartbeat every heartbeat_ms (3s in CI),
  // so with the request timeout the pending read below would fail.
  const db = new Client(`${HOST}:${PORT}`, {
    ...(API_KEY ? { apiKey: API_KEY } : {}),
    timeoutMs: 1000,
  });
  const coll = db.collection(uniqueCollection());
  const stream = coll.watch();
  let iterator: AsyncIterator<unknown> | null = null;
  let firstSettled = false;
  try {
    await coll.insertOne({ _seed: true }); // Watch requires an existing collection.
    iterator = stream[Symbol.asyncIterator]();
    // The first next() opens the subscription, then blocks on the stream.
    const first = iterator.next();
    first.then(
      () => (firstSettled = true),
      () => (firstSettled = true), // settled below; never leave it unhandled
    );

    const IDLE = Symbol("idle");
    let early: unknown;
    try {
      // Several request-timeout windows with no events; only heartbeats flow.
      early = await settleWithin(first, 5000, IDLE);
    } catch (e) {
      if (e instanceof AuthError) {
        t.skip(`change streams not enabled on server: ${e.message}`);
        return;
      }
      throw new Error(`idle watch died: ${String(e)}`);
    }
    assert.equal(early, IDLE, "idle watch produced an event before any write");

    const docId = await coll.insertOne({ n: 1 });
    const res = await settleWithin(first, 10_000, null);
    assert.ok(res, "no change event within 10s of the write");
    assert.equal(res.done, false);
    const ev = res.value as { op: string; docId: string; document: Record<string, unknown> | null };
    assert.equal(ev.op, "upsert");
    assert.equal(ev.docId, docId);
    assert.equal(ev.document?.n, 1);
  } finally {
    if (firstSettled) {
      // Generator is parked at a yield; return() runs its cleanup.
      await iterator?.return?.();
    } else {
      // A read is still pending; return() would queue behind it. Drop the
      // socket instead so the failure path does not wait for the idle timeout.
      await stream.close();
    }
    db.close();
  }
});

File: clients/go/live_test.go
package zydecodb

import (
	"context"
	"fmt"
	"net"
	"os"
	"sort"
	"testing"
	"time"
)

// Integration tests against a live ZydecoDB server. Set ZYDECODB_TEST_HOST /
// ZYDECODB_TEST_PORT (and optionally ZYDECODB_TEST_API_KEY) to point at a
// running server; the suite is skipped when the server is unreachable, so
// `go test ./...` stays green offline.

func testAddr() string {
	host := os.Getenv("ZYDECODB_TEST_HOST")
	if host == "" {
		host = "127.0.0.1"
	}
	port := os.Getenv("ZYDECODB_TEST_PORT")
	if port == "" {
		port = "9470"
	}
	return net.JoinHostPort(host, port)
}

func liveClient(t *testing.T) *Client {
	t.Helper()
	addr := testAddr()
	c, err := net.DialTimeout("tcp", addr, time.Second)
	if err != nil {
		t.Skipf("no ZydecoDB server at %s: %v", addr, err)
	}
	_ = c.Close()
	var opts []Option
	if key := os.Getenv("ZYDECODB_TEST_API_KEY"); key != "" {
		opts = append(opts, WithAPIKey(key))
	}
	return NewClient(addr, opts...)
}

func uniqueCollection() string {
	return fmt.Sprintf("gotest_%d", time.Now().UnixNano())
}

func TestLivePing(t *testing.T) {
	c := liveClient(t)
	defer c.Close()
	if err := c.Ping(context.Background()); err != nil {
		t.Fatalf("ping: %v", err)
	}
}

func TestKV(t *testing.T) {
	db := liveClient(t)
	defer db.Close()
	ctx := context.Background()

	key := []byte(fmt.Sprintf("testkv_%d", time.Now().UnixNano()))
	val := []byte("hello kv")

	// Get missing
	res, err := db.Get(ctx, key)
	if err != nil {
		t.Fatal(err)
	}
	if res != nil {
		t.Fatalf("expected nil, got %q", res)
	}

	// Put
	seq, err := db.Put(ctx, key, val, 0)
	if err != nil {
		t.Fatal(err)
	}
	if seq == 0 {
		t.Fatal("expected positive seq")
	}

	// Get
	res, err = db.Get(ctx, key)
	if err != nil {
		t.Fatal(err)
	}
	if string(res) != "hello kv" {
		t.Fatalf("got %q, want hello kv", res)
	}

	// Delete
	existed, err := db.Delete(ctx, key)
	if err != nil {
		t.Fatal(err)
	}
	if !existed {
		t.Fatal("expected true")
	}
	existed, err = db.Delete(ctx, key)
	if err != nil {
		t.Fatal(err)
	}
	if existed {
		t.Fatal("expected false")
	}

	// Put with TTL (expired)
	expired := uint64(time.Now().Add(-time.Hour).UnixMilli())
	_, err = db.Put(ctx, key, val, expired)
	if err != nil {
		t.Fatal(err)
	}
	res, err = db.Get(ctx, key)
	if err != nil {
		t.Fatal(err)
	}
	if res != nil {
		t.Fatalf("expected nil for expired key, got %q", res)
	}
}

func TestLiveInsertFindUpdateDelete(t *testing.T) {
	c := liveClient(t)
	defer c.Close()
	ctx := context.Background()
	coll := c.Collection(uniqueCollection())

	if _, err := coll.CreateIndex(ctx, []string{"age"}, false, 0); err != nil {
		t.Fatalf("create index: %v", err)
	}
	ids, err := coll.InsertMany(ctx, []Document{
		{"name": "Ada", "age": 30, "city": "London"},
		{"name": "Bo", "age": 25, "city": "NOLA"},
		{"name": "Cy", "age": 40, "city": "NOLA"},
	})
	if err != nil {
		t.Fatalf("insert many: %v", err)
	}
	if len(ids) != 3 {
		t.Fatalf("expected 3 ids, got %d", len(ids))
	}

	got, err := coll.Find(ctx, Document{"age": Document{"$gte": 30}}, QueryOptions{
		Sort: []SortKey{{Field: "age", Ascending: true}},
	})
	if err != nil {
		t.Fatalf("find: %v", err)
	}
	var names []string
	for _, d := range got {
		names = append(names, d["name"].(string))
	}
	if len(names) != 2 || names[0] != "Ada" || names[1] != "Cy" {
		t.Fatalf("unexpected find result: %v", names)
	}

	res, err := coll.UpdateOne(ctx, Document{"name": "Bo"}, Document{"$inc": Document{"age": 10}}, false, false)
	if err != nil {
		t.Fatalf("update: %v", err)
	}
	if res.Matched != 1 || res.Modified != 1 {
		t.Fatalf("unexpected update result: %+v", res)
	}

	cnt, err := coll.CountDocuments(ctx, nil)
	if err != nil {
		t.Fatalf("count: %v", err)
	}
	if cnt != 3 {
		t.Fatalf("expected count 3, got %d", cnt)
	}

	cities, err := coll.Distinct(ctx, "city", nil)
	if err != nil {
		t.Fatalf("distinct: %v", err)
	}
	strs := make([]string, 0, len(cities))
	for _, v := range cities {
		strs = append(strs, v.(string))
	}
	sort.Strings(strs)
	if len(strs) != 2 || strs[0] != "London" || strs[1] != "NOLA" {
		t.Fatalf("unexpected distinct: %v", strs)
	}

	deleted, err := coll.DeleteMany(ctx, Document{"city": "NOLA"}, false)
	if err != nil {
		t.Fatalf("delete many: %v", err)
	}
	if deleted != 2 {
		t.Fatalf("expected 2 deleted, got %d", deleted)
	}
}

func TestLiveUniqueIndexConflict(t *testing.T) {
	c := liveClient(t)
	defer c.Close()
	ctx := context.Background()
	coll := c.Collection(uniqueCollection())

	if _, err := coll.CreateIndex(ctx, []string{"email"}, true, 0); err != nil {
		t.Fatalf("create unique index: %v", err)
	}
	if _, err := coll.InsertOne(ctx, Document{"email": "a@b.com"}, false, 0); err != nil {
		t.Fatalf("first insert: %v", err)
	}
	_, err := coll.InsertOne(ctx, Document{"email": "a@b.com"}, false, 0)
	if err == nil || !IsConflict(err) {
		t.Fatalf("expected conflict error, got %v", err)
	}
}

func TestLiveUpsertSetOnInsert(t *testing.T) {
	c := liveClient(t)
	defer c.Close()
	ctx := context.Background()
	coll := c.Collection(uniqueCollection())

	miss, err := coll.UpdateOne(ctx,
		Document{"email": "soi@example.com"},
		Document{
			"$set":         Document{"email": "soi@example.com", "n": 1.0},
			"$setOnInsert": Document{"created": true},
		},
		false, true,
	)
	if err != nil {
		t.Fatalf("upsert miss: %v", err)
	}
	if miss.Matched != 0 || miss.Modified != 0 || miss.UpsertedID == "" {
		t.Fatalf("unexpected miss result: %+v", miss)
	}
	doc, err := coll.FindOne(ctx, Document{"email": "soi@example.com"}, QueryOptions{})
	if err != nil {
		t.Fatalf("find after upsert: %v", err)
	}
	if doc["created"] != true || doc["n"].(float64) != 1 {
		t.Fatalf("unexpected doc after insert: %#v", doc)
	}

	hit, err := coll.UpdateOne(ctx,
		Document{"email": "soi@example.com"},
		Document{
			"$set":         Document{"n": 2.0},
			"$setOnInsert": Document{"created": false, "extra": 1.0},
		},
		false, true,
	)
	if err != nil {
		t.Fatalf("upsert hit: %v", err)
	}
	if hit.Matched != 1 || hit.Modified != 1 || hit.UpsertedID != "" {
		t.Fatalf("unexpected hit result: %+v", hit)
	}
	doc, err = coll.FindOne(ctx, Document{"email": "soi@example.com"}, QueryOptions{})
	if err != nil {
		t.Fatalf("find after hit: %v", err)
	}
	if doc["n"].(float64) != 2 || doc["created"] != true {
		t.Fatalf("setOnInsert must be ignored on hit: %#v", doc)
	}
	if _, ok := doc["extra"]; ok {
		t.Fatalf("extra must not appear on hit: %#v", doc)
	}
}

func TestLiveOptimisticConcurrency(t *testing.T) {
	c := liveClient(t)
	defer c.Close()
	ctx := context.Background()
	coll := c.Collection(uniqueCollection())

	id, err := coll.InsertOne(ctx, Document{"n": 1.0}, false, 0)
	if err != nil {
		t.Fatalf("insert: %v", err)
	}
	got, err := coll.GetWithRevision(ctx, id)
	if err != nil || got.Doc == nil {
		t.Fatalf("get with revision: %v %#v", err, got)
	}
	if got.Revision == 0 {
		t.Fatal("expected non-zero revision")
	}
	newRev, err := coll.ReplaceOneIfMatch(ctx, id, Document{"n": 2.0}, false, got.Revision, 0)
	if err != nil {
		t.Fatalf("replace if match: %v", err)
	}
	if newRev <= got.Revision {
		t.Fatalf("expected newer revision, got %d after %d", newRev, got.Revision)
	}
	_, err = coll.ReplaceOneIfMatch(ctx, id, Document{"n": 3.0}, false, got.Revision, 0)
	if !IsConflict(err) {
		t.Fatalf("expected conflict on stale replace, got %v", err)
	}
	after, err := coll.UpdateByIDIfMatch(ctx, id, Document{"$inc": Document{"n": 1.0}}, false, newRev)
	if err != nil {
		t.Fatalf("update if match: %v", err)
	}
	if after <= newRev {
		t.Fatalf("expected newer revision after update, got %d", after)
	}
}

func TestLiveBoundedTransaction(t *testing.T) {
	c := liveClient(t)
	defer c.Close()
	ctx := context.Background()
	collName := uniqueCollection()
	coll := c.Collection(collName)
	if _, err := coll.InsertOne(ctx, Document{"_seed": true}, false, 0); err != nil {
		t.Fatalf("seed collection: %v", err)
	}

	kvKey := []byte("tx-go-" + collName)
	seq, err := c.WithTransaction(ctx, func(tx *Tx) error {
		if err := tx.Put(ctx, kvKey, []byte("v1"), 0); err != nil {
			return err
		}
		if err := tx.PutDocument(ctx, collName, "u1", Document{"n": 1.0}); err != nil {
			return err
		}
		got, err := tx.Get(ctx, kvKey)
		if err != nil {
			return err
		}
		if string(got) != "v1" {
			t.Fatalf("read-your-writes: got %q", got)
		}
		return nil
	})
	if err != nil {
		t.Fatalf("transaction: %v", err)
	}
	if seq == 0 {
		t.Fatal("expected non-zero commit seq")
	}
	val, err := c.Get(ctx, kvKey)
	if err != nil || string(val) != "v1" {
		t.Fatalf("committed kv: %v %q", err, val)
	}

	// Rollback path
	tx, err := c.BeginTx(ctx)
	if err != nil {
		t.Fatalf("begin: %v", err)
	}
	if err := tx.Put(ctx, []byte("tx-rollback"), []byte("nope"), 0); err != nil {
		t.Fatalf("stage: %v", err)
	}
	if err := tx.Rollback(ctx); err != nil {
		t.Fatalf("rollback: %v", err)
	}
	gone, err := c.Get(ctx, []byte("tx-rollback"))
	if err != nil || gone != nil {
		t.Fatalf("rollback must leave nothing: %v %q", err, gone)
	}
}
